src/core/ir/eval.rs

```diff
@@ -12,10 +12,10 @@ use crate::core::serde_value_ext::ValueExt;
 
 impl IR {
     #[tracing::instrument(skip_all, fields(otel.name = %self), err)]
-    pub fn eval<'a, Ctx>(
+    pub fn eval<'a, 'b, Ctx>(
         &'a self,
-        ctx: &'a mut EvalContext<'a, Ctx>,
-    ) -> Pin<Box<impl Future<Output = Result<ConstValue, Error>> + 'a>>
+        ctx: &'b mut EvalContext<'a, Ctx>,
+    ) -> Pin<Box<dyn Future<Output = Result<ConstValue, Error>> + Send + 'b>>
```

Comment:
IMPORTANT: This is introducing a dynamic dispatch. Please revert.

Reply:
I'm trying to resolve the issue about not having the Mutex in EvaluationContext here.

Since we removed Eval trait from main I'm struggling to fix the code now because of the errors rustc generates. I raised the discussion https://users.rust-lang.org/t/rpitit-allows-more-flexible-code-in-comparison-with-raw-rpit-in-inherit-impl/113417 about this. 

In short I can't specify impl here without breaking usage later because of the mutable reference to ctx. dyn works right now. Adding Eval trait will also solve this.